### ISSUE TITLE
Problem with `\multicolumn` in first cell of a tabularx tcolorboxes

### ISSUE THREAD

**Issue by nignasi** [T-F-S/tcolorbox#60]
When a `tabularx` option is used within tcolorbox, it's not possible to define a `\multicolumn` command on the first cell (just after tcolorbox parameters). It works when used on any other cell in first row or any other row.

This problem has been reported in https://tex.stackexchange.com/questions/351170/tcolorbox-tabularx-first-line-with-a-multicolumn-does-not-work-in-tcolorbox-en. From David Carlisle comments it seems that `\multicolumn` should be the first thing in cell and he suspects that tcolorbox introduces some other character before.

Following code shows that there's no problem with `\multicolumn` except when it's typed on first cell, but the solution proposed in TeX.SX fails because the supressed extra row is not completely erased (I suspect due to some extra space in rows).

As usual, thank you for your help.

Ignasi

```
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[table]{xcolor}
\usepackage{array,tabularx}
\usepackage[most]{tcolorbox}

\begin{document}

\begin{tcolorbox}[enhanced,
	title=A table, coltitle=white, 
	clip upper,
    tabularx*={\rowcolors{1}{red!30}{white}}{>{\centering\arraybackslash}X%
              >{\centering\arraybackslash}X}]
A & B \\ 
1 & 2 \\ 
3 & 4
\end{tcolorbox}

\begin{tcolorbox}[enhanced,
	title=A table, coltitle=white, 
	clip upper,
    tabularx*={\rowcolors{1}{red!30}{white}}{>{\centering\arraybackslash}X%
              >{\centering\arraybackslash}X}]
1 & 2 \\ 
\multicolumn{2}{c}{AB}\\
3 & 4
\end{tcolorbox}

%Following code fails. 
%Uncomenting the line with `noalign` partially solves the problem.
%Code compiles, but colors are inverted and a partially colored lines is on top.
\begin{tcolorbox}[enhanced,
	title=A table, coltitle=white, 
	clip upper,
    tabularx*={\rowcolors{1}{red!30}{white}}{>{\centering\arraybackslash}X%
              >{\centering\arraybackslash}X}]
%\\ \noalign{\vspace{-\normalbaselineskip}} 
\multicolumn{2}{c}{AB}\\
1 & 2 \\ 
3 & 4
\end{tcolorbox}
\end{document}
```

**Comment by T-F-S** [comment]
I can confirm the problem. Also, David Carlisle is right (as always) that there is some invisible stuff before the first cell. Here, it is an `\ignorespaces`.

As a short hack for now, I can give you an option `hackfortabularx` which removes the `\ignorespaces`. The following should work:

```latex
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[table]{xcolor}
\usepackage{array,tabularx}
\usepackage[most]{tcolorbox}


\makeatletter
\tcbset{%
  hackfortabularx/.code={%
    \def\tcb@saveupperbox{%
    \begin{tcb@savebox}{\tcb@upperbox}{\tcb@w@upper}\kvtcb@fontupper\kvtcb@halignupper\kvtcb@before@upper}}
}
\makeatother

\begin{document}

\begin{tcolorbox}[enhanced,
	title=A table, coltitle=white,
	clip upper,
    tabularx*={\rowcolors{1}{red!30}{white}}{>{\centering\arraybackslash}X%
              >{\centering\arraybackslash}X}]
A & B \\ 
1 & 2 \\ 
3 & 4
\end{tcolorbox}

\begin{tcolorbox}[enhanced,
	title=A table, coltitle=white,
	clip upper,
    tabularx*={\rowcolors{1}{red!30}{white}}{>{\centering\arraybackslash}X%
              >{\centering\arraybackslash}X}]
1 & 2 \\ 
\multicolumn{2}{c}{AB}\\
3 & 4
\end{tcolorbox}

%Following code fails.
%Uncomenting the line with `noalign` partially solves the problem.
%Code compiles, but colors are inverted and a partially colored lines is on top.
\begin{tcolorbox}[enhanced,breakable,
	title=A table, coltitle=white,
	clip upper,
	hackfortabularx,
    tabularx*={\rowcolors{1}{red!30}{white}}{>{\centering\arraybackslash}X%
              >{\centering\arraybackslash}X}]
%\\ \noalign{\vspace{-\normalbaselineskip}}
\multicolumn{2}{c}{AB}\\
1 & 2 \\ 
3 & 4
\end{tcolorbox}
\end{document}
```

I will see, if I can incorporate this directly into the `tabularx` option without breaking other things. Mainly, an inner `tcolorbox` inside `tabularx` will inherit the hack ... does anyone use a `tcolorbox` inside `tabularx` inside a `tcolorbox`? One never knows for sure...

**Comment by T-F-S** [comment]
The problem should be solved with version 4.20. The patch is no longer needeed.

**Comment by nignasi** [comment]
thank you
